content/rsf-spec/index.md

```diff
@@ -0,0 +1,263 @@
+TSEP             = ":"    ; time separator
+TZ               = "Z"    ; timezone
+```
+
+### Media type
+
+The current media type is `application/uk-gov-rsf`. It should change to
+`application/vnd.rsf` to align with [RFC6838](https://tools.ietf.org/html/rfc6838).
+
+### REST API
+
+TODO
+
+The current implementation uses `GET /download-rsf`. The main issue with that
+is that diverges from the rest of the API where serialisation is expressed
+either via suffix or via media type. The problem with using the same approach,
+say `GET /register.rsf` is that we are not providing the same information when
+querying `GET /register.json`.
+
+What is a good name for a resource that represents the whole raw database?
+
+```
+# RSF
+
+GET /{db resource}.rsf
+
+GET /{db resource}
+Accept: application/vnd.rsf
+
+
+# JSON Lines — Hypothetical
+
+GET /{db resource}.jsonl
+
+GET /{db resource}.jsonl
+Accept: application/x-ndjson
+```
+
+
+### Commands
+
+#### <a id="assert-root-hash-command">`assert-root-hash` command</a>
+
+Asserts that the provided root hash is the same as the one computed from the
+current entry log as defined in the [Digital Proofs][digital-proofs]
+specification.
+
+##### Arguments
+
+1. The `hash` of the root of the tree.
+
+For example, the empty root hash:
+
+```
+assert-root-hash	sha-256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+```
+
+#### <a id="add-item-command">`add-item` command</a>
+
+Adds a new [Item resource][item-res] to the register. It will require an
+[`append-entry` command](#append-entry-command) to make it visible to users.
+
+##### Arguments
+
+1. The [canonical representation][canon-rep] of the item.
+
+For exeample:
+
+```
+add-item	{"country":"GB","name":"United Kingdom","official-name":"The United Kingdom of Great Britain and Northern Ireland"}
+```
+
+#### <a id="append-entry-command">`append-entry` command</a>
+
+Appends a new [Entry resource][entry-res] to the register.
+
+##### Arguments
+
+1. The `type` of the entry determines if the entry belongs to the data log
+   (`user`) or to the metadata log (`system`).
+2. The `key` of the entry. The primary key field is the field with the same
+   name as the register.
+3. The `timestamp` of the entry. This is the time at which the entry was
+   appended to the register.
+4. The `hash` of the item for which the entry was appended. This is the
+   [sha-256 hash of the item][canon-rep].
+
+For example:
+
+```
+append-entry	user	GB	2010-11-12T13:14:15Z	sha-256:08bef0039a4f0fb52f3a5ce4b97d7927bf159bc254b8881c45d95945617237f6
+```
+
+
+### Validation rules
+
+A RSF list of commands is expected to conform to the following rules:
+
+* [Commands](#commands) are executed in order of appearance, top to bottom.
+* Entries are numbered in sequence in order of appearance starting with 1 if
+  the register is empty, otherwise incrementing on the latest entry number
+  found in the register.
```

Comment:
Is there any way to tell whether an RSF file represents a complete register, or only appends to a register?

Reply:
The only way as far as I know is via `assert-root-hash`. In a situation where the RSF has no assertions you wouldn't know.

Reply:
In that case I think it would be useful to require an assert-root-hash in the first line of every file, for the sake of reassembling a register from its patches.

Reply:
Should we specify that this currently does not include `system` entries (though it should do).

Reply:
By this I mean that entry numbers are computed independently for `user` and `system` currently (as they are separate logs on the inside).

Reply:
Yes, it needs to explain the current implementation exactly.